homeassistant/components/foscam/switch.py

```diff
@@ -22,63 +110,87 @@ async def async_setup_entry(
 
-        self._attr_unique_id = f"{config_entry.entry_id}_sleep_switch"
-        self._attr_translation_key = "sleep_switch"
-        self._attr_has_entity_name = True
+        self.entity_description = description
+        self._attr_unique_id = f"{entry_id}_{description.key}"
 
-        self.is_asleep = self.coordinator.data["is_asleep"]["status"]
+        if description.key == "is_asleep":
+            self._state = self.coordinator.data.is_asleep["status"]
+        else:
+            self._state = getattr(self.coordinator.data, self.entity_description.key)
 
     @property
-    def is_on(self):
-        """Return true if camera is asleep."""
-        return self.is_asleep
+    def is_on(self) -> bool:
+        """Return the state of the switch."""
+        return self._state
 
     async def async_turn_off(self, **kwargs: Any) -> None:
-        """Wake camera."""
-        LOGGER.debug("Wake camera")
-
-        ret, _ = await self.hass.async_add_executor_job(
-            self.coordinator.session.wake_up
-        )
-
-        if ret != 0:
-            raise HomeAssistantError(f"Error waking up: {ret}")
-
+        """Turn off the entity."""
+        if self.entity_description.key == "is_open_ir":
+            if self.entity_description.set_ir_config_auto_close:
+                self.hass.async_add_executor_job(
+                    self.entity_description.set_ir_config_auto_close,
+                    self.coordinator.session,
+                )
+        if self.entity_description.turn_off_fn is not None:
+            self.hass.async_add_executor_job(
+                self.entity_description.turn_off_fn, self.coordinator.session
+            )
         await self.coordinator.async_request_refresh()
 
     async def async_turn_on(self, **kwargs: Any) -> None:
-        """But camera is sleep."""
-        LOGGER.debug("Sleep camera")
-
-        ret, _ = await self.hass.async_add_executor_job(self.coordinator.session.sleep)
-
-        if ret != 0:
-            raise HomeAssistantError(f"Error sleeping: {ret}")
-
+        """Turn on the entity."""
+        if self.entity_description.key == "is_open_ir":
+            if self.entity_description.set_ir_config_auto:
+                self.hass.async_add_executor_job(
+                    self.entity_description.set_ir_config_auto, self.coordinator.session
+                )
+        self.hass.async_add_executor_job(
+            self.entity_description.turn_on_fn, self.coordinator.session
+        )
         await self.coordinator.async_request_refresh()
 
     @callback
     def _handle_coordinator_update(self) -> None:
         """Handle updated data from the coordinator."""
-
-        self.is_asleep = self.coordinator.data["is_asleep"]["status"]
-
+        if self.entity_description.key == "is_asleep":
+            self._state = self.coordinator.data.is_asleep["status"]
+        else:
+            self._state = getattr(self.coordinator.data, self.entity_description.key)
```

Comment:
Instead, let's create a `native_value_fn` callable, and have this logic in the entity description. The benefit of doing that is that we get improved type checking :)